doc/tutorials/content/voxel_grid_structure.rst

```diff
@@ -0,0 +1,238 @@
+  
+  * Used for iterating through the voxels
+  * Return any type which can be incremented and compared, e.g., iterator, index
+
+* ``bool setUp(CartesianFilter<pcl::Filter, VoxelStruct>&)``
+  
+  * Initialize the member variables for filtering
+  * Input point cloud and all settings of CartesianFilter can be accessed from the reference
+
+* ``void addPointToGrid(PointT&)``
+
+  * Add a single point to a specific voxel
+
+* ``pcl::experimental::optional<PointT> filterGrid(...)``
+
+  * Accept output from ``begin()`` and ``end()``
+  * Decide whether the iterating voxel is eligible
+  * Push a point representing the iterating voxel
+
+3. VoxelFilter
+--------------
+
+``VoxelFilter<VoxelStruct<Voxel<PointT>,PointT>>``
+
+It implements the API specific for ``VoxelGrid`` and accepts template argument ``VoxelStruct`` as the filtering logic.
+
+
+Example: Create a custom grid filter
+==========================
+
+An example is worth a throusand words, let's implement a simple grid filters easily by reusing our classes!
+
+1. Define a custom grid cell
+----------------------------
+
+This is a simple grid cell example which stores the number of points in the cell and can accumulate XYZ coordinates, 
+and it can compute the centroid by averaging.
+
+You should determine the information you need to store for your custom grid cell.
+
+.. note::
+  The following member functions are not necessary if you are not using our ``VoxelStruct``.
+
+.. code-block:: cpp
+
+  template <typename PointT>
+  struct MyVoxel {
+    // Accumulate points
+    inline void add(const PointT& pt) {
+      num_pt_++;
+      centroid_ += pt.getArray4fMap();
+    }
+
+    // Export a point to the output
+    inline PointT get() const {
+      PointT pt;
+      pt.getArray4fMap() = centroid_ / num_pt_;
+      return pt;
+    }
+
+    // Empty the voxel
+    inline void clear() {
+      num_pt_ = 0;
+      centroid_ = Eigen::Array4f::Zero();
+    }
+
+    // Return the number of added points
+    inline std::size_t size() const { return num_pt_; }
+
+  protected:
+    // Enable Eigen SSE with Array4f instead Array3f 
+    Eigen::Array4f centroid_ = Eigen::Array4f::Zero();
+    std::size_t num_pt_ = 0;
+  };
+
+.. note::
+  **Tips**:
+  If your application needs to filter some incoming point clouds repeatedly,
+  reusing the same filter object instead of creating a new one is way faster.
+  It is because cpp don't have to allocate the grid cells again and again.
+
+2. Define a custom VoxelStruct
+------------------------------
+
+This is a simple custom ``VoxelStruct`` withs a fixed grid size and fixed condition for selecting eligible voxels.
+
+We also implement a custom point hashing function ``myHash`` to select which grid cells we want to add our points to.
+(you can also use our implementation ``hashPoint<PointT>(...)`` but for simplity we don't use that here)
+
+.. note::
+  The following member functions except our custom ``myHash`` are mandatory.
+
+.. code-block:: cpp
```

Comment:
Would it be possible to instead add a CMakeLists.txt + source file and refer to the lines in that instead?